src/library/scala/collection/immutable/Map.scala

```diff
@@ -206,11 +206,22 @@ object Map extends MapFactory[Map] {
 
   def empty[K, V]: Map[K, V] = EmptyMap.asInstanceOf[Map[K, V]]
 
-  def from[K, V](it: collection.IterableOnce[(K, V)]): Map[K, V] =
+  def from[K, V](it: IterableOnce[(K, V)]): Map[K, V] =
     it match {
       case it: Iterable[_] if it.isEmpty => empty[K, V]
-      case m: Map[K, V] => m
-      case _ => (newBuilder[K, V] ++= it).result()
+      // Since IterableOnce[(K, V)] launders the variance of K,
+      // identify only our implementations which can be soundly substituted.
+      // For example, the ordering used by sorted maps would fail on widened key type. (scala/bug#12745)
+      // The following type test is not sufficient: case m: Map[K, V] => m
+      case m: HashMap[K, V]    => m
+      case m: Map1[K, V]       => m
+      case m: Map2[K, V]       => m
+      case m: Map3[K, V]       => m
+      case m: Map4[K, V]       => m
+      //case m: WithDefault[K, V] => m    // cf SortedMap.WithDefault
+      case m: SeqMap[K, V]     => SeqMap.from(it)
```

Comment:
This is problematic for Scala.js. It will cause `SeqMap` to be reachable in basically all codebases, preventing it from being dead code eliminated.

Reply:
Aha. I didn't like reaching into the subclass. That was really just for the private `SeqMap1` etc, which we can forego. Lukas made me do it! Do you mean the `SeqMap` companion, that is, I can type test for `VectorMap` here and that is OK?

Reply:
The type tests are fine. It's the instantiation via `SeqMap.from` that is problematic.

Reply:
This is fine now, thanks :)